compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -867,6 +875,7 @@ impl<'a: 'ast, 'ast, 'tcx> LateResolutionVisitor<'a, '_, 'ast, 'tcx> {
         following_seg: Option<&Segment>,
         span: Span,
         base_error: &BaseError,
+        candidates: &mut Vec<ImportSuggestion>,
```

Comment:
You need, at the very least, to add a comment above `fn suggest_typo` that describes in what scenarios the input `candidates` will be modified, and what kind of modification it will undergo.

From reading the code change here, this looks like a very blunt instrument of outright clearing the candidates in the presence of a shadowed name. 

Are you planning on doing something more sophisticated to `candidates` in the long term (e.g. to *still* suggest some new imports that are based on some small revision to the given symbol)?

If that is not the case, and all you intend is to just clear the candidate set, I think this interface is more confusing than it needs to be. I would consider something else, e.g. returning something richer than a `bool` as the return value, and then letting the caller interpret that `bool` in deciding whether to return the candidates, or to return a fresh `vec![]`.